clients/go/admin/client.go

```diff
@@ -54,105 +81,195 @@ func GetAdditionalAdminClientConfigOptions(cfg Config) []grpc.DialOption {
+	}, nil
+}
+
+// InitializeAuthMetadataClient creates a new anonymously Auth Metadata Service client.
+func InitializeAuthMetadataClient(ctx context.Context, cfg *Config) (client service.AuthMetadataServiceClient, err error) {
+	onceAuthMetadata.Do(func() {
+		authMetadataConnection, err = NewAdminConnection(ctx, cfg)
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize admin connection. Error: %w", err)
+	}
+
+	return service.NewAuthMetadataServiceClient(authMetadataConnection), nil
+}
+
+func NewAdminConnection(_ context.Context, cfg *Config, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	if opts == nil {
+		// Initialize opts list to the potential number of options we will add. Initialization optimizes memory
+		// allocation.
+		opts = make([]grpc.DialOption, 0, 5)
+	}
 
 	if cfg.UseInsecureConnection {
 		opts = append(opts, grpc.WithInsecure())
 	} else {
 		// TODO: as of Go 1.11.4, this is not supported on Windows. https://github.com/golang/go/issues/16736
 		creds := credentials.NewClientTLSFromCert(nil, "")
 		opts = append(opts, grpc.WithTransportCredentials(creds))
-		if cfg.UseAuth {
-			logger.Infof(ctx, "Instantiating a token source to authenticate against Admin, ID: %s", cfg.ClientID)
-			jwtDialOption, err := getAuthenticationDialOption(ctx, cfg)
-			if err != nil {
-				return nil, err
-			}
-			opts = append(opts, jwtDialOption)
-		}
 	}
 
 	opts = append(opts, GetAdditionalAdminClientConfigOptions(cfg)...)
+
 	return grpc.Dial(cfg.Endpoint.String(), opts...)
 }
 
 // Create an AdminClient with a shared Admin connection for the process
-func InitializeAdminClient(ctx context.Context, cfg Config) service.AdminServiceClient {
+// Deprecated: Please use initializeClients instead.
+func InitializeAdminClient(ctx context.Context, cfg *Config, opts ...grpc.DialOption) service.AdminServiceClient {
+	set, err := initializeClients(ctx, cfg, nil, opts...)
+	if err != nil {
+		logger.Panicf(ctx, "Failed to initialize client. Error: %v", err)
+		return nil
+	}
+
+	return set.AdminClient()
+}
+
+// initializeClients creates an AdminClient and AuthServiceClient with a shared Admin connection for the process.
+// Note that if called with different cfg/dialoptions, it will not
+func initializeClients(ctx context.Context, cfg *Config, tokenCache pkce.TokenCache, opts ...grpc.DialOption) (*Clientset, error) {
 	once.Do(func() {
-		var err error
-		adminConnection, err = NewAdminConnection(ctx, cfg)
+		authMetadataClient, err := InitializeAuthMetadataClient(ctx, cfg)
+		if err != nil {
+			logger.Panicf(ctx, "failed to initialize Auth Metadata Client. Error: %v", err)
+		}
+
+		// If auth is enabled, this call will return the required information to use to authenticate, otherwise,
+		// start the client without authentication.
+		opt, err := getAuthenticationDialOption(ctx, cfg, tokenCache, authMetadataClient)
+		if err != nil {
+			logger.Warnf(ctx, "Starting an unauthenticated client because: %v", err)
```

Comment:
we always warn then if auth is not enabled?

Reply:
yeah... an incentive to enable auth? thoughts?

Reply:
as long as it's intentional, sounds good :)